aptos-move/aptos-aggregator/src/bounded_math.rs

```diff
@@ -0,0 +1,215 @@
+// Copyright © Aptos Foundation
+
+use aptos_logger::error;
+use move_binary_format::errors::{PartialVMError, PartialVMResult};
+use move_core_types::vm_status::StatusCode;
+
+// When bounded math operation overflows
+// Generally means addition exceeded limit.
+pub(crate) const EBOUND_OVERFLOW: u64 = 0x02_0001;
+
+/// When bounded math operation underflows
+/// Generally means subtraction went below 0.
+pub(crate) const EBOUND_UNDERFLOW: u64 = 0x02_0002;
+
+/// When updating the aggregator start value (due to read operations
+/// or at the end of the transaction), we realize that mistakenly raised
+/// an overflow in one of the previus try_add operation.
+pub(crate) const EEXPECTED_OVERFLOW: u64 = 0x02_0003;
+
+/// When updating the aggregator start value (due to read operations
+/// or at the end of the transaction), we realize that mistakenly raised
+/// an underflow in one of the previus try_sub operation.
+pub(crate) const EEXPECTED_UNDERFLOW: u64 = 0x02_0004;
+
+pub(crate) const ECODE_INVARIANT_BROKEN: u64 = 0x02_0005;
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum BoundedMathError {
+    Overflow,
+    Underflow,
+}
+
+/// Error for delta application. Can be used by delta partial functions
+/// to return descriptive error messages and an appropriate error code.
+pub(crate) fn abort_error(message: impl ToString, code: u64) -> PartialVMError {
+    PartialVMError::new(StatusCode::ABORTED)
+        .with_message(message.to_string())
+        .with_sub_status(code)
+}
+
+impl From<BoundedMathError> for PartialVMError {
+    fn from(err: BoundedMathError) -> Self {
+        match err {
+            BoundedMathError::Overflow => abort_error("Overflow", EBOUND_OVERFLOW),
+            BoundedMathError::Underflow => abort_error("Underflow", EBOUND_UNDERFLOW),
+        }
+    }
+}
+
+pub type BoundedMathResult<T> = ::std::result::Result<T, BoundedMathError>;
+
+pub fn ok_overflow<T>(value: BoundedMathResult<T>) -> BoundedMathResult<Option<T>> {
+    match value {
+        Ok(value) => Ok(Some(value)),
+        Err(BoundedMathError::Overflow) => Ok(None),
+        Err(BoundedMathError::Underflow) => Err(BoundedMathError::Underflow),
+    }
+}
+
+pub fn ok_underflow<T>(value: BoundedMathResult<T>) -> BoundedMathResult<Option<T>> {
+    match value {
+        Ok(value) => Ok(Some(value)),
+        Err(BoundedMathError::Overflow) => Err(BoundedMathError::Overflow),
+        Err(BoundedMathError::Underflow) => Ok(None),
+    }
+}
+
+pub fn expect_ok<T>(value: BoundedMathResult<T>) -> PartialVMResult<T> {
```

Comment:
I thought this method is never used in the speculative context, but I do see that it just map the bounded math error. In this case, error! here is wrong, as it will log errors for cases that happened during speculative execution (i.e. with wrong speculative value). If we want to log, what we need to use is the speculative logging crate, but that needs transaction index. So it'd have to be up in the chain of calls, as long as we can figure out an aggregator error happened, we can log way up the call chain? @igor-aptos @vusirikala 
  
If this was intended to only be used when the operations are supposed to succeed, then I'm not sure about the uses below (I see uses in try_add, try_sub methods). I think it's the same question for me to understand the difference of how ECODE_INVARIANT_BROKEN is used.

Reply:
this is called for cases that should never happen, if code is valid, even in speculative context.

For example, we are maintaining that speculative_start_value + delta is always in the [0, max_value] range. We check that before updating either of the two. 
If that is broken at any point, ECODE_INVARIANT_BROKEN will be thrown.

Reply:
I think it was the use at line e.g. 357 that confused me, adding the new input, but now I think I figured, ok_overflow actually is okay with an overflow, and then we check we are okay :)

Reply:
expect_ok(ok_overflow()) asserts there is no underflow